### PR TITLE
C++: Represent temporary object initialization in AST and IR

### PR DESCRIPTION
**Pull request by dbartol**
Fixes github/codeql-c-analysis-team#157, which is the first part of github/codeql-c-analysis-team#154.

This is the `github/codeql` side of the work. The dbscheme is updated to add a new expression kind, `@temp_init`, which is treated as a conversion and thus not in the parent-child tree in the AST. These expressions are represented by the new `TemporaryObjectExpr` (naming bikeshed still open), which has the initializer as its `getExpr()`. Very few queries should notice the change, since most don't look at conversions anyway, and those that do generally care specifically about numeric conversions anyway.

The IR changes to support these expressions went pretty smoothly. The code is very similar to how we handle `ThrowExpr`, where we also have an initializer that initializes a temporary object.

This fixes all remaining consistency failures in the IR test directory, as well as a bunch of similar consistency failures in the `syntax-zoo`.

Most of the test diffs are just in tests that dump lists of expected expressions, and now include the occasional `TemporaryObjectExpr`. However, the diffs in `library-tests/dataflow/taint-tests` bear close examination (@geoffw0?):
- This change fixes dozens of false negatives.
- Fixes several ODR violations that were making it difficult to dump ASTs or IR.
- Subtle tweaks to our mock STL implementation to make the behavior consistent with the standard. The biggest of these is to make `std::make_pair("foo", "blah")` return a `std::pair<const char*, const char*>`, rather than `std::pair<const char[4], const char[5]>` (the argument types are supposed to undergo "decay").
- Several models have been updated to use `isParamDeref()` instead of `isParam()` as appropriate.

I also made some changes to how local taint flow is computed in `DefaultTaintTracking.qll`. The main change is that model flow now goes from `Operand`->`Instruction`, rather than `Instruction`->`Instruction`. This fixes a couple dozen cases of missed taint flow in `library-tests/dataflow/DefaultTaintTracking`.

To do:
- [x] No-op upgrade script
- [x] Diffs: https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1532//artifact/changes.html

### CODE REVIEW

File: cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/Instruction.qll
Comment (dbartol) [inline cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/Instruction.qll:807]: All these diffs in all copies of `Instruction.qll` should go away once #4502 is merged.

**Comment by rdmarsh2** [comment]
The QL code in the first commit looks good to me, but I've only skimmed the test changes and I'm not sure I've reviewed all the relevant QL changes. Could you do a rebase onto main to get rid of the test/display changes that were cherry-picked into separate PRs?

**Comment by dbartol** [comment]
I've just merged from `main`, so I believe any remaining diffs are from the PR, not the other dump formatting PRs.

File: cpp/ql/test/library-tests/ir/ir/raw_ir.expected
Comment (rdmarsh2) [inline cpp/ql/test/library-tests/ir/ir/raw_ir.expected:7582]: Shouldn't the `this` operand to a `CallInstruction` have a pointer, reference, or `glaval<...>` type?

File: cpp/ql/test/library-tests/ir/ir/raw_ir.expected
Comment (rdmarsh2) [inline cpp/ql/test/library-tests/ir/ir/raw_ir.expected:7573]: Shouldn't the `this` operand to a `CallInstruction` have a pointer, reference, or `glval<...>` type? I think it also breaks the `AddressOperand`s for the side effects.

There's the same issue with the method calls on lines 1374, 1385, 1386, 1397 and 1398.

File: cpp/ql/src/semmle/code/cpp/ir/implementation/raw/internal/TranslatedElement.qll
Comment (rdmarsh2) [inline cpp/ql/src/semmle/code/cpp/ir/implementation/raw/internal/TranslatedElement.qll:270]: I think we need to exclude the qualifier here, and instead use the `VariableAddressInstruction` generated for the `TemporaryObjectExpr` as the `this` operand for the call. See the comment in the tests.

File: cpp/ql/src/semmle/code/cpp/exprs/Cast.qll
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/exprs/Cast.qll:855]: The GNU binary ternary operator also seems to create these expressions. Is that intended? Looking at `question_mark_colon/cfg.expected`, it looks like it's created even for trivial types.

File: cpp/ql/src/semmle/code/cpp/models/implementations/MemberFunction.qll
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/models/implementations/MemberFunction.qll:42]: Surely `input.isParameter(0)` was wrong to begin with, and one of the data-flow libraries should be adjusted so we only have to write `input.isParameterDeref(0)` here. Or? Not that it has to happen in this PR.

Cc @geoffw0

File: cpp/ql/src/semmle/code/cpp/models/implementations/StdPair.qll
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/models/implementations/StdPair.qll:23]:
```suggestion
    this.getParameter(0).getUnspecifiedType().(ReferenceType).getBaseType()
```

There should be no need for a second `getUnspecifiedType` call -- the first one is recursive and goes all the way down.

File: cpp/ql/test/library-tests/dataflow/DefaultTaintTracking/stl.cpp
Comment (jbj) [inline cpp/ql/test/library-tests/dataflow/DefaultTaintTracking/stl.cpp:25]: There's no way `basic_string` is implemented like this. If the data-flow modelling only works with this mock implementation, then the data-flow modelling needs to be fixed. We've committed to supporting `basic_string`, so this very important constructor needs to work.

File: cpp/ql/src/semmle/code/cpp/ir/implementation/raw/internal/TranslatedElement.qll
Comment (dbartol) [reply cpp/ql/src/semmle/code/cpp/ir/implementation/raw/internal/TranslatedElement.qll:270]: Ah, good catch, although the bug isn't in this predicate. The issue is that C++ 98 allowed the qualifier of a member access to be an rvalue, so EDG marks the temporary object created as the qualifier to be a prvalue. More recent standards changed the wording to that if the qualifier is a prvalue, it undergoes temporary object materialization to turn it into a glvalue, so that the converted qualifier is always a glvalue. EDG presumably preserves the load for compatibility, but we should ignore it. This predicate won't re-add the load, because `call.getAnArgument()` does not include the qualifier.

The same problem exists with field accesses on prvalues. I'll have a fix and updated tests soon.

**Comment by jbj** [comment]
You've added a lot of code to the IR generation to bridge between how the IR likes to see the AST and how the AST looks in the database, especially with the last commit about member accesses on prvalues. Would it be an advantage to move any of that code to the layer that's in between the IR and the database: the QL AST classes?

File: cpp/ql/src/semmle/code/cpp/ir/implementation/raw/internal/TranslatedElement.qll
Comment (dbartol) [reply cpp/ql/src/semmle/code/cpp/ir/implementation/raw/internal/TranslatedElement.qll:322]: Fixed.

File: cpp/ql/test/library-tests/question_mark_colon/cfg.expected
Comment (dbartol) [reply cpp/ql/test/library-tests/question_mark_colon/cfg.expected]: Sorry, that was an old diff that I hadn't re-accepted

File: cpp/ql/test/library-tests/dataflow/taint-tests/movableclass.cpp
Comment (geoffw0) [inline cpp/ql/test/library-tests/dataflow/taint-tests/movableclass.cpp:23]: Were these (`&&`) just not necessary to get move semantics all the way through lines 59 and 60?

File: cpp/ql/test/library-tests/dataflow/taint-tests/test_ir.expected
Comment (geoffw0) [inline cpp/ql/test/library-tests/dataflow/taint-tests/test_ir.expected:70]: The `// [FALSE POSITIVE]` label on this line should be removed.

**Review by geoffw0 — COMMENTED** [review]
Taint test changes LGTM.  The IR catches a lot of new results, and a few new false positives but similar ones to AST taint (and in most cases on lines we expect to fail, e.g. due to an insufficient model of `swap` and `clear`).

File: cpp/ql/src/semmle/code/cpp/models/implementations/MemberFunction.qll
Comment (geoffw0) [reply cpp/ql/src/semmle/code/cpp/models/implementations/MemberFunction.qll:42]: I agree, we should just delete `input.isParameter(0)` (or if that breaks any of the AST tests then ask me to do it in a follow-up PR instead).

File: cpp/ql/test/library-tests/ir/ir/raw_ir.expected
Comment (dbartol) [reply cpp/ql/test/library-tests/ir/ir/raw_ir.expected]: These should all be fixed now, with a consistency check added.

File: cpp/ql/test/library-tests/dataflow/DefaultTaintTracking/stl.cpp
Comment (dbartol) [reply cpp/ql/test/library-tests/dataflow/DefaultTaintTracking/stl.cpp]: Oops, another leftover experiment. I've fixed this for real by fixing taint flow to qualifier objects in `DefaultTaintTracking`.

File: cpp/ql/test/library-tests/dataflow/taint-tests/movableclass.cpp
Comment (dbartol) [reply cpp/ql/test/library-tests/dataflow/taint-tests/movableclass.cpp:23]: It's not so much that they weren't necessary, as that they weren't well-defined behavior. If you return a reference to a temporary by rvalue reference, it's still a dangling reference. To return a movable object, you just return it by value. The construction of the return value will use the move constructor (although in this case, the constructor is probably elided entirely). When you then use the return value in the caller, that return value is an xvalue, so it will get move semantics as well.

File: cpp/ql/src/semmle/code/cpp/ir/dataflow/DefaultTaintTracking.qll
Comment (MathiasVP) [inline cpp/ql/src/semmle/code/cpp/ir/dataflow/DefaultTaintTracking.qll:219]: To be consistent with how we interleave `instruction -> operand` steps in dataflow we should probably convert this to pure `instruction -> operand` and `operand -> instruction` flow at some point once this PR has been merged. But I'm fine with these changes as they are now!

Comment (dbartol) [reply cpp/ql/src/semmle/code/cpp/ir/dataflow/DefaultTaintTracking.qll:219]: Yeah, I just didn't want to change more than I already had to.

**Comment by dbartol** [comment]
Very few diffs. The `implicit-bitfield-downcast` is a side effect of fixing a "one `Expr` with multiple immediate `Conversion`s" issue. The correct fix is suggested in https://github.com/github/codeql-c-analysis-team/issues/180.

**Comment by dbartol** [comment]
JDK diff looks like a true positive. A call to a pure constructor wasn't treated as being a `PureExprInVoidContext` because it had a `void` return type. With the addition of the temporary object into which the object is constructed, the return type is now the type of the class, so we'll treat it the same as a call to a pure function that returns that class.

**Comment by dbartol** [comment]
php-src diff did not repo locally.

**Comment by dbartol** [comment]
Diffs look good. Perf is nearly unchanged. This is ready for final sign-off.

**Review by geoffw0 — COMMENTED** [review]
I've just reviewed the code changes as best I can and they look good to me.  I remember checking the taint test result changes from this PR quite thoroughly before and was very happy with them as well.

[A comment](https://github.com/github/codeql/pull/4432/files#r508454092]) by @jbj appears to have gone missing from the front page of the PR.  Apart from that, it appears all concerns have been addressed.

I'm keen to merge this today.  We've certainly all had plenty of opportunity to review at this point.  I've started SAMATE runs (https://jenkins.internal.semmle.com/job/Security/job/SAMATE/job/SAMATE-cpp/548/ and https://jenkins.internal.semmle.com/job/Security/job/SAMATE/job/SAMATE-cpp2/31/) on this branch, mostly for my own information - they will take ~18 hours to complete so lets not wait for them.

File: cpp/ql/src/semmle/code/cpp/ir/dataflow/DefaultTaintTracking.qll
Comment (jbj) [reply cpp/ql/src/semmle/code/cpp/ir/dataflow/DefaultTaintTracking.qll:369]: I think I've understood why. We don't want it to look as if there is a definite dereference. That would cause FPs in queries like "null value dereferenced here".

I'm happy to delete/replace the comment, but it doesn't have to be on this PR.

**Comment by jbj** [comment]
Please don't merge this PR until the `rc/1.26` branch has been created. At the time of writing this, it's waiting for a submodule bump to finish testing.

**Comment by geoffw0** [comment]
> Please don't merge this PR until the rc/1.26 branch has been created.

I foolishly assumed that had happened yesterday as planned.  Is there somewhere I can check for sure, or can you post here when it's done?

**Comment by jbj** [comment]
> I foolishly assumed that had happened yesterday as planned. Is there somewhere I can check for sure, or can you post here when it's done?

It's done, and we can move forward. But this branch can only be merged together with its internal PR, and the internal PR has both a test failure and a submodule conflict.

File: cpp/ql/src/semmle/code/cpp/ir/dataflow/DefaultTaintTracking.qll
Comment (dbartol) [reply cpp/ql/src/semmle/code/cpp/ir/dataflow/DefaultTaintTracking.qll:369]: I think so. Read side effects tend to assume that we might read anything that can be reached from the pointer via pointer arithmetic.

**Comment by dbartol** [comment]
>You've added a lot of code to the IR generation to bridge between how the IR likes to see the AST and how the AST looks in the database, especially with the last commit about member accesses on prvalues. Would it be an advantage to move any of that code to the layer that's in between the IR and the database: the QL AST classes?

The fixups that I'm doing involve introducing new temporary objects where there were none before, so I'm not sure how I'd do that in the AST layer without drastically changing how we implement the AST. I've considered fixing this all up in the extractor, but if it's primarily for the IR's benefit, I think I'd prefer to add more QL to the IR construction than more C++ to the extractor.

**Review by jbj — APPROVED** [review]

File: cpp/ql/src/semmle/code/cpp/models/implementations/MemberFunction.qll
Comment (geoffw0) [reply cpp/ql/src/semmle/code/cpp/models/implementations/MemberFunction.qll:42]: Opened https://github.com/github/codeql/pull/4642 .

**Pull request merged**
